### PR TITLE
Improve OkHttp Docs

### LINKED ISSUE #6565: [Proposal] Improve OkHttp Documentation
This issue serves as a first step in proposing changes to the OkHttp Documentation found under https://square.github.io/okhttp

## The reasons
While the docs do look nice do they miss out on a lot of different features that are provided by the 'Material for MkDocs' Theme (Which I call just 'Material' throughout this issue for simplicity reasons) and PyMdown Extensions (Also known as Pymdownx).

## Changes
I propose the following main changes to the documentation.

### Dark theme
Material offers a simple setting for using dark theme called slate:
```yaml
theme:
  palette:
    scheme: slate
```

And let's be honest: This bright theme is not good for the eyes to look at and a dark theme would be a lot better to use.

Note, that the dark theme was addes in a later 5.x version of Material, so upgrading should be recommended.

### Using tabs
Material offers the usage of tabs.
This allows you to keep pages organized in tabs, which are visible at the top of the page.

Adding it only requires this in the mkdocs.yml
```yaml
theme:
  features:
  - navigation.tabs
```

As mentioned allows this to group pages together under a single name, which makes it look better in comparison to just having all pages listed on the side.

As an example can you group and add the `changelog_3x.md`, `changelog_2x.md`, `changelog_1x.md` and `changelog.md` under the `Change Log` section like this:

```yaml
nav:
# ...
- 'Change Logs': # I suggest renaming this to Change Logs
  - 'Latest': changelog.md
  - 'Version 3.x': changelog_3x.md
  - 'Version 2.x': changelog_2x.md
  - 'Version 1.x': changelog_1x.md
```

### General reordering of pages
Next proposal is to reorder and group pages together into their own separate sections.
You already have `4.x` as a separate group, which considering all other pages being separated seems somewhat odd and inconsistent.

I propose the following Folder structure to use:
```
docs/
  │
  ├─ 4.x/
  │  ├─ mockwebserver/
  │  │  └─ okhttp3.mockwebserver/
  │  │     └─ index.md
  │  ├─ okhttp/
  │  │  └─ okhttp3/
  │  │     └─ index.md
  │  ├─ okhttp-brotli/
  │  │  └─ okhttp3.brotli/
  │  │     └─ index.md
  │  ├─ okhttp-dnsoverhttps/
  │  │  └─ okhttp3.dnsoverhttps/
  │  │     └─ index.md
  │  ├─ okhttp-logging-interceptor/
  │  │  └─ okhttp3.logging/
  │  │     └─ index.md
  │  ├─ okhttp-sse/
  │  │  └─ okhttp3.sse/
  │  │     └─ index.md
  │  ├─ okhttp-tls/
  │  │  └─ okhttp3.tls/
  │  │     └─ index.md
  │  ├─ okhttp-urlconnection/
  │  │  └─ okhttp3/
  │  │     └─ index.md
  │  └─ upgrading_to_okhttp_4.md
  ├─ changelogs/
  │  ├─ changelog.md
  │  ├─ changelog_3x.md
  │  ├─ changelog_2x.md
  │  ├─ changelog_1.md
  │  └─ tls_configuration_history.md
  ├─ contribute/
  │  ├─ code_of_conduct.md
  │  └─ contributing.md
  ├─ features/
  │  ├─ caching.md
  │  ├─ calls.md
  │  ├─ concurrency.md
  │  ├─ connections.md
  │  ├─ debug_logging.md
  │  ├─ events.md
  │  ├─ https.md
  │  ├─ inteceptors.md
  │  └─ r8_proguard.md
  ├─ security/
  │  ├─ security.md
  │  └─ security_providers.md
  ├─ index.md
  ├─ recipes.md
  ├─ releasing.md
  └─ works_with_okhttp.md
```

The benefits of such a structure would be a more organized look and feel.
Note that this folder structure would also be mostly mirrored in the navigation as follows:
```yaml
nav:
nav:
  - 'Overview':
    - 'Introduction': index.md
    - 'Stack Overflow ⏏': https://stackoverflow.com/questions/tagged/okhttp?sort=active
  - 'Features':
    - 'Caching': features/caching.md
    - 'Calls': features/calls.md
    - 'Concurrency': features/concurrency.md
    - 'Connections': features/connections.md
    - 'Debug Logging': features/debug_logging.md
    - 'Events': features/events.md
    - 'HTTPS': features/https.md
    - 'Interceptors': features/interceptors.md
    - 'R8 / ProGuard': features/r8_proguard.md
  - 'Recipes': recipes.md
  - 'Security':
    - 'Security Policy': security/security.md
    - 'Security Providers': security/security_providers.md
  - 'Works with OkHttp': works_with_okhttp.md
  - '4.x API':
    - 'Upgrading to OkHttp 4': 4.x/upgrading_to_okhttp_4.md
    - 'okhttp': 4.x/okhttp/okhttp3/index.md
    - 'brotli': 4.x/okhttp-brotli/okhttp3.brotli/index.md
    - 'dnsoverhttps': 4.x/okhttp-dnsoverhttps/okhttp3.dnsoverhttps/index.md
    - 'logging-interceptor': 4.x/okhttp-logging-interceptor/okhttp3.logging/index.md
    - 'sse': 4.x/okhttp-sse/okhttp3.sse/index.md
    - 'tls': 4.x/okhttp-tls/okhttp3.tls/index.md
    - 'urlconnection': 4.x/okhttp-urlconnection/okhttp3/index.md
    - 'mockwebserver': 4.x/mockwebserver/okhttp3.mockwebserver/index.md
  - '3.12.x API':
    - 'okhttp ⏏': https://square.github.io/okhttp/3.x/okhttp/
    - 'dnsoverhttps ⏏': https://square.github.io/okhttp/3.x/okhttp-dnsoverhttps/
    - 'logging-interceptor ⏏': https://square.github.io/okhttp/3.x/logging-interceptor/
    - 'sse ⏏': https://square.github.io/okhttp/3.x/okhttp-sse/
    - 'tls ⏏': https://square.github.io/okhttp/3.x/okhttp-tls/
    - 'urlconnection ⏏': https://square.github.io/okhttp/3.x/okhttp-urlconnection/
    - 'mockwebserver ⏏': https://square.github.io/okhttp/3.x/mockwebserver/
  - 'Change Logs':
    - 'Latest': changelogs/changelog.md
    - 'Version 3.x': changelogs/changelog_3x.md
    - 'Version 2.x': changelogs/changelog_2x.md
    - 'Version 1.x': changelogs/changelog_1x.md
    - 'TLS History': changelogs/tls_configuration_history.md
  - 'Contributing':
    - 'Contribute': contribute/contributing.md
    - 'Code of Conduct': contribute/code_of_conduct.md
```

## Usage of other pymdownx extensions
I believe using the Emojis extension of pymdownx together with the MaterialX emoji option, which is part of Material and enables the usage of SVG icons from Fontawesome, Material Design and Octicons, would be beneficial for the documentation.

Right now is the main benefit/feature I imagined, which is displaying something like Fontawesome's `external-link-alt` icon in the navigation tree on the left, not supported.

~~I [opened an issue](https://github.com/squidfunk/mkdocs-material/issues/2318) on Material proposing such a change, as it would look better to use such an icon to indicate external pages, rather than using ⏏.~~
Unfortunately is this not doable for extensions and themes.

But perhaps there would also be other places where the usage of either emojis or icons could be beneficial to use.

## Live Preview
All the changes I discussed above can be found on a live-preview I made over here: https://andre601.ch/okhttp

## Notes
I used the following things for this:
- MkDocs 1.1
- Material for MkDocs 6.2.8
- pymdown-extensions 7.x
- mkdocs-redirects 1.0.1

I used `pip install -U mkdocs-material` to upgrade Material and all its dependencies and `pip install mkdocs-redirects` to download the redirects plugin used.

Another note:
With exception of a placeholder index file for the main page do some other pages not exist due to the nature of OkHttp's Documentation structure.

### PR DESCRIPTION
**Pull request by Andre601**
Updates the complete docs in an attempt to make them more readable and organized.
This should not cause conflicts with other programs/systems (i.e. Javadoc generation, but I won't guarantee this, so testing it first is highly recommendet.

Please see #6565 for full context and actual major changes.

closes #6565

### CODE REVIEW

**Comment by swankjesse** [comment]
I’d like to preserve the existing URLs, even though these URLs are slightly nicer. Keeping other peoples’ links working is nice, and there’s like a million things that link to our docs.

**Comment by Andre601** [comment]
Maybe re-add the old files, but the file's content informs about the move, rather than having a duplicate

**Comment by swankjesse** [comment]
Can we keep the new structure but the old paths?

**Comment by Andre601** [comment]
Not possible to my knowledge

**Comment by Andre601** [comment]
I now made following Changes:

- Added MkDocs Redirect plugin. This plugin will create redirects for pages, allowing the old paths to now redirect to the new ones.
  All old paths have been added as a redirect to the new ones now.
  Note that in order for it to work locally will you need to install the plugin first:
  ```
  pip install mkdocs-redirects
  ```
- Fixed invalid links to other pages and images. Maybe consider to use full URLs for at least the images?

This change should now make the old links work again, which was tested. Note that I didn't re-add the placeholder files this time with the exception of the index one, so the live-preview will have a few broken pages.

**Comment by swankjesse** [comment]
Nice find on the redirect plugin!

**Comment by Andre601** [comment]
I've made some new changes:

- Moved images and CSS into a dedicated `assets` directory (More clean).
- Added a `.icons/brands` directory in `assets` containing two SVG icons (kotlin and Java logo). They are from simpleicons and a license was added too.
- These SVG icons are used in the Java and Kotlin tabs to make them a bit more pretty. The Recipes section is probs best example of it.

Note that there seems to be a minor issue with the pymdownx.emoji extension where it won't render the SVG icons (`:brands-kotlin:` and `:brands-java:` when not all pages are available, so it probs requires these docs here to be generated last.

**Comment by Andre601** [comment]
Updated `deploy_website.sh`.

It should now copy `changelog.md` and `contributing.md` to their new respective folders.
I hope this will work as expected, so that mkdocs won't have issues with missing or wrong placed files.

**Comment by yschimke** [comment]
https://www.andre601.ch/okhttp/

image

**Comment by Andre601** [comment]
> https://www.andre601.ch/okhttp/
> 
> image

Is something with that?

**Comment by yschimke** [comment]
> As always is the live-version available to see the changes in action: https://andre601.ch/okhttp

Showing "Hi" at the live-version page.

I assumed this was a bug.

**Comment by Andre601** [comment]
> > As always is the live-version available to see the changes in action: https://andre601.ch/okhttp
> 
> Showing "Hi" at the live-version page.
> 
> I assumed this was a bug.

I mentioned this in the linked Issue:
> Another note:
With exception of a placeholder index file for the main page do some other pages not exist due to the nature of OkHttp's Documentation structure.

**Comment by Andre601** [comment]
Any update on this?
It has been open for almost a year now and would like to know if there is anything I need to change, update, etc.

**Review by jj-crypto — APPROVED** [review]

**Comment by Andre601** [comment]
I'm considering doing some extra work on this stuff here.

- First would I go and add a theme switch, so that people could choose if they want to use the light or dark theme (which one is used by default would depend on their general system settings)
- Next would I add some extra CSS which I found from other docs that in the end would add a `external link` icon for every page that is not part of the default docs (Would essentially lead to an external page)

It is a bit sad that this PR hasn't been looked into at all for months now. I do understand that the lib itself has the main priority, but I cannot understand why there isn't a simple response here about the current status... It makes this PR feel abandoned by you.

**Comment by yschimke** [comment]
That looks really good, I think we should go with it.  cc @swankjesse who knows more about the docs process.

I would definitely say, let's get this in before starting more work on this, so perhaps another PR that would build on this.

One nit: is there a broken link for the latest changelog? https://www.andre601.ch/okhttp/changelogs/changelog/

**Comment by Andre601** [comment]
> One nit: is there a broken link for the latest changelog? https://www.andre601.ch/okhttp/changelogs/changelog/

I mentioned this in the linked Issue:
> [With exception of a placeholder index file for the main page do some other pages not exist due to the nature of OkHttp's Documentation structure.](https://github.com/square/okhttp/issues/6565#issue-812922863)

The changelog file is a 404 because it doesn't seem to exist in the default docs directory. At least when I forked and updated the stuff was this the case.
I'm not sure if this is due to some of the automatic build setup you use that this is the case, but I didn't want to add a file to not break stuff.

**Comment by Andre601** [comment]
> I'm considering doing some extra work on this stuff here.
> 
> * First would I go and add a theme switch, so that people could choose if they want to use the light or dark theme (which one is used by default would depend on their general system settings)
> * Next would I add some extra CSS which I found from other docs that in the end would add a `external link` icon for every page that is not part of the default docs (Would essentially lead to an external page)

I now pushed the mentioned changes which should also be visible in a few minutes on the live-demo at https://andre601.ch/okhttp

**Comment by swankjesse** [comment]
This is great. I want to apologize for not giving this any oxygen for a full year; that’s on me for spreading myself thin and should not detract from my enthusiasm for this feature.

File: docs/recipes.md [hunk 1]
```diff
@@ -559,370 +564,365 @@ It is an error to have multiple caches accessing the same cache directory simult
...
+          System.out.println("Response 2 response:          " + response2);
+          System.out.println("Response 2 cache response:    " + response2.cacheResponse());
+          System.out.println("Response 2 network response:  " + response2.networkResponse());
+        }
+    
+        System.out.println("Response 2 equals Response 1? " + response1Body.equals(response2Body));
+      }
+    ```
+
+To prevent a response from using the cache, use [`CacheControl.FORCE_NETWORK`](http://square.github.io/okhttp/4.x/okhttp/okhttp3/-cache-control/-f-o-r-c-e_-n-e-t-w-o-r-k/). To prevent it from using the network, use [`CacheControl.FORCE_CACHE`](http://square.github.io/okhttp/4.x/okhttp/okhttp3/-cache-control/-f-o-r-c-e_-c-a-c-h-e/). Be warned: if you use `FORCE_CACHE` and the response requires the network, OkHttp will return a `504 Unsatisfiable Request` response.
```
Comment (swankjesse) [inline docs/recipes.md:644]: For follow-up, let’s change these URLs back to HTTPS

**Review by swankjesse — APPROVED** [review]
What’s up with the `okhttp-hpacktests` submodule change?

**Comment by swankjesse** [comment]
Revert the off-topic hpacktests change: https://github.com/square/okhttp/pull/7056

**Comment by swankjesse** [comment]
A few notes on landing this:

 * Search broke. Material for Mkdocs only includes the search plugin if no other plugins are added. I fixed this.
 * The ‘external link’ icon isn’t working and I’d like to get rid of it. Tracking here: https://github.com/square/okhttp/issues/7060
 
 Thanks again!

**Comment by Andre601** [comment]
V3 docs also seem broken

**Comment by yschimke** [comment]
Finally, the long arctic winter of Light Mode is over!

**Pull request merged**